### PR TITLE
ARROW-14315: [C++][Gandiva] Implement BROUND function

### PR DESCRIPTION
**Pull request by augustoasilva**
Returns the rounded BIGINT value of a using HALF_EVEN rounding mode. Also known as Gaussian rounding or bankers' rounding.

### CODE REVIEW

File: cpp/src/gandiva/precompiled/extended_math_ops.cc [hunk 1]
```diff
@@ -236,6 +237,25 @@ gdv_int64 round_int64(gdv_int64 num) { return num; }
 ROUND_DECIMAL(float32)
 ROUND_DECIMAL(float64)
 
+// rounds the number to the nearest integer
+#define BROUND_DECIMAL(TYPE)                       \
+  FORCE_INLINE                                     \
+  gdv_##TYPE bround_##TYPE(gdv_##TYPE num) {       \
+    const gdv_float64 round_num = round(num);      \
+    const gdv_float64 dif_num = round_num - num;   \
+    if ((dif_num != 0.5f) && (dif_num != -0.5f)) { \
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/extended_math_ops.cc:246]: Should be 0.5f or 0.5d since dif_num is a double?

File: cpp/src/gandiva/precompiled/extended_math_ops.cc [hunk 2]
```diff
@@ -236,6 +237,25 @@ gdv_int64 round_int64(gdv_int64 num) { return num; }
 ROUND_DECIMAL(float32)
 ROUND_DECIMAL(float64)
 
+// rounds the number to the nearest integer
+#define BROUND_DECIMAL(TYPE)                       \
+  FORCE_INLINE                                     \
+  gdv_##TYPE bround_##TYPE(gdv_##TYPE num) {       \
+    const gdv_float64 round_num = round(num);      \
+    const gdv_float64 dif_num = round_num - num;   \
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/extended_math_ops.cc:245]: typo - please use diff_num

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc [hunk 3]
```diff
@@ -120,6 +120,50 @@ TEST(TestExtendedMathOps, TestRoundDecimal) {
...
+  VerifyFuzzyEquals(bround_float32(2.5f), 2);
+  VerifyFuzzyEquals(bround_float32(3.5f), 4);
+  VerifyFuzzyEquals(bround_float32(-2.5f), -2);
+  VerifyFuzzyEquals(bround_float32(-3.5f), -4);
+  VerifyFuzzyEquals(bround_float32(1.4999999f), 1);
+  VerifyFuzzyEquals(bround_float64(2.5), 2);
+  VerifyFuzzyEquals(bround_float64(3.5), 4);
+  VerifyFuzzyEquals(bround_float64(-2.5), -2);
+  VerifyFuzzyEquals(bround_float64(-3.5), -4);
+  VerifyFuzzyEquals(bround_float64(1.4999999), 1);
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/extended_math_ops_test.cc:152]: Please add a test for bround_float64(1.50001) = 2

File: cpp/src/gandiva/function_registry_arithmetic.cc [hunk 4]
```diff
@@ -107,6 +107,12 @@ std::vector<NativeFunction> GetArithmeticFunctionRegistry() {
       BINARY_GENERIC_SAFE_NULL_IF_NULL(round, {}, int32, int32, int32),
       BINARY_GENERIC_SAFE_NULL_IF_NULL(round, {}, int64, int32, int64),
 
+      // bround functions
+      UNARY_SAFE_NULL_IF_NULL(bround, {}, float32, float32),
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/function_registry_arithmetic.cc:111]: Yes it does not have, I added float for convenience. But I removed it.

File: cpp/src/gandiva/precompiled/extended_math_ops.cc (hunk 1 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/extended_math_ops.cc:246]: Yes, it should, it will not have a support for float too.

File: cpp/src/gandiva/precompiled/extended_math_ops.cc (hunk 2 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/extended_math_ops.cc:245]: Fixed

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc (hunk 3 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/extended_math_ops_test.cc:152]: Test added

File: cpp/src/gandiva/function_registry_arithmetic.cc [hunk 5]
```diff
@@ -107,6 +107,13 @@ std::vector<NativeFunction> GetArithmeticFunctionRegistry() {
       BINARY_GENERIC_SAFE_NULL_IF_NULL(round, {}, int32, int32, int32),
       BINARY_GENERIC_SAFE_NULL_IF_NULL(round, {}, int64, int32, int64),
 
+      // bround functions
+      NativeFunction("bround", {}, DataTypeVector{float64()}, float64(),
+                     kResultNullIfNull, "bround_float64"),
+
+      NativeFunction("bround", {}, DataTypeVector{float64(), int32()}, float64(),
```
Comment (vvellanki) [inline cpp/src/gandiva/function_registry_arithmetic.cc:114]: Lets not add this in this PR. Please open a separate ticket for this and create a separate PR.

File: cpp/src/gandiva/precompiled/extended_math_ops.cc [hunk 6]
```diff
@@ -249,6 +265,26 @@ ROUND_DECIMAL(float64)
...
 ROUND_DECIMAL_TO_SCALE(float64)
 
+// rounds the number to the given scale
+FORCE_INLINE
+gdv_float64 bround_float64_int32(gdv_float64 num, gdv_int32 out_scale) {
+  if (out_scale < 0) {
+    out_scale = out_scale * (-1);
+  }
+  gdv_float64 scale_multiplier = get_scale_multiplier(out_scale);
+  gdv_float64 to_round = num * scale_multiplier;
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/extended_math_ops.cc:275]: This can cause an overflow. This implementation has to be thought through - please do it in a separate PR

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc [hunk 7]
```diff
@@ -120,6 +120,46 @@ TEST(TestExtendedMathOps, TestRoundDecimal) {
...
 
+TEST(TestExtendedMathOps, TestBRoundDecimal) {
+  EXPECT_DOUBLE_EQ(bround_float64(0.0), 0);
+  EXPECT_DOUBLE_EQ(bround_float64(2.5), 2);
+  EXPECT_DOUBLE_EQ(bround_float64(3.5), 4);
+  EXPECT_DOUBLE_EQ(bround_float64(-2.5), -2);
+  EXPECT_DOUBLE_EQ(bround_float64(-3.5), -4);
+  EXPECT_DOUBLE_EQ(bround_float64(1.4999999), 1);
+  EXPECT_EQ(std::signbit(bround_float64(0)), 0);
+  EXPECT_EQ(std::signbit(bround_float64(0)), 0);
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/extended_math_ops_test.cc:131]: This is same as the test in line 130

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc [hunk 8]
```diff
@@ -120,6 +120,46 @@ TEST(TestExtendedMathOps, TestRoundDecimal) {
...
+  EXPECT_DOUBLE_EQ(bround_float64_int32(6.225, 2), 6.22);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(8.335, 2), 8.34);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(-8.25, 1), -8.2);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(-8.35, 1), -8.4);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(-6.225, 2), -6.22);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(-8.335, 2), -8.34);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(-6.225, -2), -6.22);
+  EXPECT_DOUBLE_EQ(bround_float64_int32(8.335, -2), 8.34);
+
+  VerifyFuzzyEquals(bround_float64(2.5), 2);
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/extended_math_ops_test.cc:143]: What is the difference between VerifyFuzzyEquals and EXPECT_DOUBLE_EQ?

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc (hunk 7 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/extended_math_ops_test.cc:131]: My mistake, removed it

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc (hunk 8 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/extended_math_ops_test.cc:143]: VerifyFuzzyEquals has a tolefrance to be almost equal, not exactly the same number. Should I remove VerifyFuzzyEquals test cases?

File: cpp/src/gandiva/precompiled/extended_math_ops.cc (hunk 6 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/extended_math_ops.cc]: Ok, I will open a new PR when as soon as this one is merged

File: cpp/src/gandiva/function_registry_arithmetic.cc (hunk 5 above)
Comment (augustoasilva) [reply cpp/src/gandiva/function_registry_arithmetic.cc]: Ok, I will open a new PR when as soon as this one is merged

**Review by vvellanki — APPROVED** [review]

**Comment by pravindra** [comment]
@augustoasilva there are failures reported by clang-format with this PR. Can you pls fix those ?

https://github.com/apache/arrow/runs/4161625801?check_suite_focus=true

**Comment by augustoasilva** [comment]
@pravindra I fixed them
